protocol-spec/generic.md

```diff
@@ -134,5 +142,197 @@
+**Flow Diagram:**
+
+**Serialization Example:**
+
+```json
+[
+  {
+    "VibrateCmd": {
+      "Id": 1,
+      "DeviceIndex": 0,
+      "Speeds": [
+        {
+          "Index": 0,
+          "Speed": 0.5
+        },
+        {
+          "Index": 1,
+          "Speed": 1
+        }
+      ]
+    }
+  }
+]
+```
+
+## LinearCmd
+
+**Description:** Causes a toy that supports linear movement to reposition it's linear actuators at a certain speeds. In order to abstract the dynamic ranges (both speed and movement) of different toys, the values are sent as a dictionary of linear actuator indexes against objects encapsulating floats with a range of \[0.0-1.0\] for speed and position.
+
+**Introduced In Version:** 1
+
+**Message Version:** 1
+
+**Fields:**
+
+* _Id_ \(unsigned int\): Message Id
+* _DeviceIndex_ \(unsigned int\): Index of device
+* _Vectors_ \(array\): Linear actuator speeds and positions
+  * _Index_ \(unsigned int\): Index of linear actuator
+  * _Speed_ \(double\): Movement speed with a range of \[0.0-1.0\]
+  * _Position_ \(double\): Target position with a range of \[0.0-1.0\]
+
+**Expected Response:**
+
+* Ok message with matching Id on successful request.
+* Error message on value or message error.
+
+**Flow Diagram:**
+
+
+**Serialization Example:**
+
+```json
+[
+  {
+    "LinearCmd": {
+      "Id": 1,
+      "DeviceIndex": 0,
+      "Vectors": [
+        {
+          "Index": 0,
+          "Speed": 0.5,
+          "Position": 0.3
+        },
+        {
+          "Index": 1,
+          "Speed": 1,
+          "Posiion": 0.8
+        }
+      ]
+    }
+  }
+]
+```
+
+
+
+## RotateCmd
+
+**Description:** Causes a toy that supports rotation to run specific rotation motors at a certain speeds in specified directions. In order to abstract the dynamic range of different toys, the values are sent as a dictionary of rotation motor indexes against objects encapsulating the speed as a float with a range of \[0.0-1.0\] and the direction as a boolean (true being clockwise). **Note:** clockwise may be subjective.
+
+**Introduced In Version:** 1
+
+**Message Version:** 1
+
+**Fields:**
+
+* _Id_ \(unsigned int\): Message Id
+* _DeviceIndex_ \(unsigned int\): Index of device
+* _Speeds_ \(array\): Rotation speeds
+  * _Index_ \(unsigned int\): Index of rotation motor
+  * _Speed_ \(double\): Rotation speed with a range of \[0.0-1.0\]
+  * _Clockwise_ \(boolean\): Direction of rotation
+
+**Expected Response:**
+
+* Ok message with matching Id on successful request.
+* Error message on value or message error.
+
+**Flow Diagram:**
 
+![img](singlemotorvibratecmd_diagram.svg)
+
+**Serialization Example:**
 
+```json
+[
+  {
+    "RotationCmd": {
+      "Id": 1,
+      "DeviceIndex": 0,
+      "Speeds": [
+        {
+          "Index": 0,
+          "Speed": 0.5,
+          "Clockwise": true
+        },
+        {
+          "Index": 1,
+          "Speed": 1,
+          "Clockwise": false
+        }
+      ]
+    }
+  }
+]
+```
+
+## ConstrictCmd
```

Comment:
I'm hesitant to add this message at the moment. We only have one toy that does this, and it's not used much.

Reply:
What about the RealTouch? Didn't the mouth have a constrictor too?
I failed to find a better name for this, since I want it to apply to inflatables/expandables too

Reply:
Yeah, not that it was horribly useful, but the Real Touch did have constriction. Not saying we shouldn't have the message, just that we should possibly figure out more places where it's useful first, to make sure we're covering all the bases.

Reply:
Not that I know of any right now, but I suspect they'll be devices like inflatable buttplugs or that Indiegogo insertable with the worm gear expansion system under a silicon skin.
I guess it could be a pressure command?